src/runtime/action/index.ts

```diff
@@ -17,26 +17,35 @@
  *
  * Docs: https://svelte.dev/docs#template-syntax-element-directives-use-action
  */
-export interface ActionReturn<Parameter = any> {
-	update?: (parameter: Parameter) => void;
-	destroy?: () => void;
-}
+export type ActionReturn<Parameter = any> = void extends Parameter
+	? {
+		destroy?: () => void;
+	}
+	: {
+		update?: (parameter: Parameter) => void;
+		destroy?: () => void;
+	}
 
 /**
  * Actions are functions that are called when an element is created.
  * You can use this interface to type such actions.
  * The following example defines an action that only works on `<div>` elements
  * and optionally accepts a parameter which it has a default value for:
  * ```ts
- * export const myAction: Action<HTMLDivElement, { someProperty: boolean }> = (node, param = { someProperty: true }) => {
+ * export const myAction: Action<HTMLDivElement, undefined | { someProperty: boolean }> = (node, param = { someProperty: true }) => {
  *   // ...
  * }
  * ```
- * You can return an object with methods `update` and `destroy` from the function.
+ * If your action doesn't accept a parameter, type if as `Action<HTMLElement, void>`.
+ *
+ * You can return an object with methods `update` (if the action accepts a parameter) and `destroy` from the function.
  * See interface `ActionReturn` for more details.
  *
  * Docs: https://svelte.dev/docs#template-syntax-element-directives-use-action
  */
-export interface Action<Element = HTMLElement, Parameter = any> {
-	<Node extends Element>(node: Node, parameter?: Parameter): void | ActionReturn<Parameter>;
-}
+export type Action<Element = HTMLElement, Parameter = any> = 
+	void extends Parameter
+	? <Node extends Element>(node: Node) => void | ActionReturn<Parameter>
+	: undefined extends Parameter
+	? <Node extends Element>(node: Node, parameter?: Parameter) => void | ActionReturn<Parameter>
+	: <Node extends Element>(node: Node, parameter: Parameter) => void | ActionReturn<Parameter>;
```

Comment:
@dummdidumm will need your ts expertise over here 🙈

Reply:
I thought the same at first. Then I noticed that this will make implementing an action using the method cumbersome when you want to add a fallback parameter:

```ts
const action: Action <Element, void | {foo: boolean}>= (node, param = { foo: true }) => {
    param.foo; // TS error: Property 'foo' does not exist on type 'void | { foo: boolean; }'
}
```

But I just noticed that you can tell TS to be ok with that if you add `undefined | void`:

```ts
const action: Action <Element, void | undefined | {foo: boolean}>= (node, param = { foo: true }) => {
    param.foo; // TS is ok with this
}
```

So this means we could just leave the existing code as-is or only adjust the `ActionReturnType` interface (although I agree that this may be more hassle than it's worth). I wonder though if this is some kind of TS bug or why you can tell TS to narrow `param` to be defined when you do `void | undefined`, but not when you only do `void`.

Reply:
Interesting find. For the narrowing, I think TS is right in that `void` acts as if the function is only called with 1 argument. Maybe that's why we need both of them, `void` for the strict check in `use:action={...}` and `undefined` for inside the action implementation.

It's true that without `undefined`, both error out

![image](https://user-images.githubusercontent.com/8156777/163158212-c779ab3f-b7a4-438c-86b7-a72296874965.png)

and with `undefined`, the error correctly appears, only on the first one

![image](https://user-images.githubusercontent.com/8156777/163158343-a111499f-5d2a-4a6d-8ab1-01e251ec3f1d.png)

I checked and found that we can actually extract the `undefined` and put it directly in the Action type so it becomes `(node: Element, parameter: undefined | Parameter)`, this will also address one of my concerns of the false sense of security, while still giving the benefit of red-squiggly for users who uses the action.

I think we found a sweet spot with this where users are shown errors for required (non-optional) parameter in the actions they used, and authors would still be shown an error if they don't provide fallbacks for their parameter. If you're okay with this, I'll update the PR to reflect the desired changes.

Reply:
I created https://github.com/microsoft/TypeScript/issues/48664 in the TS repo, maybe we wait a little if they have any feedback on this.

Regarding the "false sense of security": If I remember correctly you mean that people could misuse the action and not add a parameter, even if it's required? Honestly I'm not that sure about that decision anymore. If people misuse the action, it's on them, isn't it? For example if you make an action that needs another element as input, you can't make a default parameter for that. And then authors would be forced to have an extra runtime check in there which they maybe want to avoid to make the package as small as possible.

Reply:
Good call, yeah let's wait for their feedback on the issue.

Also yes, I changed my mind and totally agree that it's their fault for misusing the action. Codebases (especially TS-based) should already be equipped with some sort of linting to mitigate these kinds of errors.

Reply:
From the response in the issue I created it sounds like we should rather not rely on this behavior - it works now, but the TS team wishes it rather wouldn't.

If we use the double conditional exports I just noticed another problem though: the `undefined extends Parameter` condition will always return true when strict mode is turned off, so the param will always be optional in non-strict mode, which is bad. Not sure how we can get to a good solution now. Maybe it's best to just leave it as is for now.. I'll try to get back to this with a fresh mind at some point.